### PR TITLE
feat: PE-20 - Added a condition lock for checking the readiness of a node to improve CPU consumption

### PR DESCRIPTION
**Pull request by Moootya**

### CODE REVIEW

File: ml_pipeline_engine/dag/graph.py [hunk 1]
```diff
@@ -1,13 +1,39 @@
...
+    source: NodeId,
+    dest: NodeId,
+    is_recurrent: bool = False,
+    is_oneof: bool = False,
+) -> DiGraph:
+    """
+    Получить связный подграф между двумя заданными узлами графа
+    """
+
+    if len(dag) == 1:
```
Comment (Afimida) [inline ml_pipeline_engine/dag/graph.py:32]: If it == 1, should it even go into this method?

File: ml_pipeline_engine/dag/manager.py [hunk 2]
```diff
@@ -22,23 +22,32 @@
...
+        default_factory=functools.partial(defaultdict, asyncio.Event),
+    )
+    condition_lock_store: t.Dict[NodeId, asyncio.Condition] = field(
+        default_factory=functools.partial(defaultdict, asyncio.Condition),
+    )
 
-    def __init__(self, node_ids: t.Iterable[NodeId]):
-        self.lock_store = {node_id: asyncio.Event() for node_id in node_ids}
+    @property
+    def cnd(self):
```
Comment (Afimida) [inline ml_pipeline_engine/dag/manager.py:45]: typehint

File: ml_pipeline_engine/dag/manager.py [hunk 3]
```diff
@@ -193,7 +205,7 @@ async def _run_node(
         if self.node_storage.exists_processed_node(node_id):
             logger.debug('Node %s has been executed. Stop new execution', node_id)
 
-            event_lock = self.lock_manager.get_lock(node_id)
+            event_lock = self._lock_manager.evn[node_id]
```
Comment (Afimida) [inline ml_pipeline_engine/dag/manager.py:208]: evn? You mean the event? Why use the short form?

File: ml_pipeline_engine/dag/graph.py [hunk 4]
```diff
@@ -1,13 +1,39 @@
...
+
+def get_connected_subgraph(
+    dag: DiGraph,
+    source: NodeId,
+    dest: NodeId,
+    is_recurrent: bool = False,
+    is_oneof: bool = False,
+) -> DiGraph:
+    """
+    Получить связный подграф между двумя заданными узлами графа
```
Comment (Afimida) [inline ml_pipeline_engine/dag/graph.py:29]: Translate doc

File: ml_pipeline_engine/dag/manager.py [hunk 5]
```diff
@@ -22,23 +22,32 @@
...
-    def __init__(self, node_ids: t.Iterable[NodeId]):
-        self.lock_store = {node_id: asyncio.Event() for node_id in node_ids}
+    @property
+    def cnd(self):
+        return self.condition_lock_store
 
-    def get_lock(self, node_id: NodeId) -> asyncio.Event:
-        self.lock_store.setdefault(node_id, asyncio.Event())
-        return self.lock_store[node_id]
+    @property
```
Comment (chaotism) [inline ml_pipeline_engine/dag/manager.py:48]: miss typehint

File: ml_pipeline_engine/dag/manager.py [hunk 6]
```diff
@@ -150,12 +162,12 @@ def _filter_node(u) -> bool:
             nx.subgraph_view(self.dag.graph, filter_edge=_filter, filter_node=_filter_node), source, dest
         )
 
-    def _get_reduced_dag_input_one_of(self, source: NodeId, dest: NodeId) -> DiGraph:
+    def _get_reduced_dag_input_one_of(self, source: NodeId, dest: NodeId, is_oneof) -> DiGraph:
```
Comment (chaotism) [inline ml_pipeline_engine/dag/manager.py:165]: miss typehint for is_oneof

File: ml_pipeline_engine/dag/graph.py (hunk 1 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/graph.py:32]: The function is common for all methods related to graph changes. Thus, I'd leave it how it is

File: ml_pipeline_engine/dag/manager.py (hunk 3 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/manager.py:208]: I wanted to decrease the length of the title of the properties. Ok, I'll change it

File: ml_pipeline_engine/dag/manager.py (hunk 2 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/manager.py]: Done

File: ml_pipeline_engine/dag/manager.py (hunk 3 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/manager.py]: Done

File: ml_pipeline_engine/dag/graph.py (hunk 4 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/graph.py]: Done

File: ml_pipeline_engine/dag/manager.py (hunk 5 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/manager.py:52]: Done

File: ml_pipeline_engine/dag/manager.py (hunk 6 above)
Comment (Moootya) [reply ml_pipeline_engine/dag/manager.py]: Done

**Review by Afimida — APPROVED** [review]

**Review by chaotism — APPROVED** [review]

**Pull request merged**